src/hid/hid_device.cc

```diff
@@ -231,45 +226,51 @@ Status HidDevice::ExchangeCbor(Command command,
   if (1 + payload.size() > kMaxDataSize) return Status::kErrInvalidLength;
   std::vector<uint8_t> send_data = {static_cast<uint8_t>(command)};
   send_data.insert(send_data.end(), payload.begin(), payload.end());
+  return SendRecvCommand(kCtapHidCbor, send_data, response_cbor, true, true,
+                         false, expect_up_check);
+}
 
-  uint8_t cmd = kCtapHidCbor;
-  OK_OR_RETURN(SendCommand(cmd, send_data));
+Status HidDevice::SendCtapHid(const std::vector<uint8_t>& payload,
+                              std::vector<uint8_t>* response) const {
+  if (payload.empty() || payload.size() - 1 > kMaxDataSize) {
+    return Status::kErrInvalidLength;
+  }
 
+  uint8_t cmd = Frame::kTypeInitMask | payload[0];
+  std::vector<uint8_t> send_data(payload.begin() + 1, payload.end());
   std::vector<uint8_t> recv_data;
-  OK_OR_RETURN(ReceiveCommand(kReceiveTimeout, &cmd, &recv_data));
-
-  // The answer might also be a keepalive.
-  bool has_sent_prompt = false;
-  while (cmd == kCtapHidKeepalive) {
-    KeepaliveStatus keepalive_response = ProcessKeepalive(recv_data);
-    if (keepalive_response == KeepaliveStatus::kStatusError)
-      return Status::kErrOther;
-    if (keepalive_response == KeepaliveStatus::kStatusUpNeeded &&
-        !has_sent_prompt) {
-      has_sent_prompt = true;
-      if (!tracker_->IsTouchPromptIgnored()) {
-        PromptUser();
-      }
-    }
-    OK_OR_RETURN(ReceiveCommand(kReceiveTimeout, &cmd, &recv_data));
+  // Cancel requires no response.
+  if (cmd == kCtapHidCancel) {
+    return SendRecvCommand(cmd, send_data, &recv_data, false, false, true,
+                           false);
   }
-
-  if (cmd != kCtapHidCbor) return Status::kErrInvalidCommand;
-  if (recv_data.empty()) return Status::kErrInvalidLength;
-
-  response_cbor->insert(response_cbor->end(), recv_data.begin() + 1,
-                        recv_data.end());
-
-  if (has_sent_prompt && !expect_up_check) {
-    tracker_->AddObservation("A prompt was sent unexpectedly.");
+  // Wink and Lock require response with empty data at success.
+  else if (cmd == kCtapHidWink || cmd == kCtapHidLock) {
+    return SendRecvCommand(cmd, send_data, &recv_data, true, false, true,
+                           false);
   }
-  if (!has_sent_prompt && expect_up_check) {
-    tracker_->AddObservation(
-        "A prompt was expected, but not performed. Sometimes it is just not "
-        "recognized if performed too fast.");
+  // Init requires non-empty response (cid) and no status code is return.
+  else if (cmd == kCtapHidInit) {
+    return SendRecvCommand(cmd, send_data, &recv_data, true, false, false,
+                           false);
+  }
+  // Ping requires response and no status code is return.
+  else if (cmd == kCtapHidPing) {
```

Comment:
The comments you need here are a sign that your design is a bit off. I feel like your are explaining the choices for your `bool`s, and as mentioned in the other review, should rather explain itself.